Fix CI

majosm opened the pull request:
**Questions for the review**:
- [ ] Is the scope and purpose of the PR clear?
  - [ ] The PR should have a description.
  - [ ] The PR should have a guide if needed (e.g., an ordering).
- [ ] Is every top-level method and class documented? Are things that should be documented actually so?
- [ ] Is the interface understandable? (I.e. can someone figure out what stuff does?) Is it well-defined?
- [ ] Does the implementation do what the docstring claims?
- [ ] Is everything that is implemented covered by tests?
- [ ] Do you see any immediate risks or performance disadvantages with the design? Example: what do interface normals attach to?

majosm committed:
install pocl 6 by default but warn about potential performance degradation issue

majosm committed:
fix pydocstyle

inducer commented on mirgecom/array_context.py:
```diff
@@ -239,6 +239,19 @@ def _check_gpu_oversubscription(actx: ArrayContext) -> None:
+    from pyopencl.characterize import get_pocl_version
+    dev = actx.queue.device
+    pocl_version = get_pocl_version(dev.platform)
+    if pocl_version[0] >= 6:
```
`pocl_version` could be None.

inducer commented on mirgecom/array_context.py:
```diff
@@ -239,6 +239,19 @@ def _check_gpu_oversubscription(actx: ArrayContext) -> None:
+        warn(
+            "Using pocl version >= 6 may degrade performance on some machines. "
+            "If running performance-critical simulations, consider downgrading "
+            "to version 5.")
```
Add link to discussion.

inducer replied:
"on some machines" too unspecific, be as specific as you can.

matthiasdiener commented:
Instead of upgrading the version for everyone, could we just upgrade it for CI?

majosm committed:
make check more specific

majosm committed:
fix mypy

matthiasdiener commented on mirgecom/array_context.py:
```diff
@@ -241,15 +241,28 @@ def _check_gpu_oversubscription(actx: ArrayContext) -> None:
+
+    # Only check if using NVIDIA
+    from pyopencl.characterize import nv_compute_capability
+    if nv_compute_capability(dev) is None:
```
Note that pocl 6 is slower on all systems, not just NVIDIA.

majosm committed:
make doublemach and doublemach_physical_av generate the same mesh

majosm commented:
@matthiasdiener This is the Mac issue I was talking about earlier: https://github.com/illinois-ceesd/mirgecom/issues/1103.

majosm committed:
temporarily xfail the failing overintegration test

need to fix loop fusion to distinguish base DOF axis from quad DOF axis for this case
to work reliably

majosm committed:
make check less specific

majosm committed:
nudge pyro eos test tolerance

majosm committed:
add issue link to xfail

majosm committed:
use non-deprecated argument in generate_gmsh

majosm commented:
Seems to be working now. The doublemach mesh stuff resulted from https://github.com/inducer/meshmode/pull/477; the two doublemach examples were previously generating slightly different meshes and "writing" them to the same file (but before that PR the output file name parameter didn't actually do anything, so it was never an issue).

matthiasdiener commented on mirgecom/array_context.py:
```diff
@@ -239,6 +239,27 @@ def _check_gpu_oversubscription(actx: ArrayContext) -> None:
+            "Using pocl version >= 6 may degrade performance when running on GPUs. "
+            "If running performance-critical simulations, consider "
+            "downgrading to version 5. See "
+            "https://github.com/illinois-ceesd/mirgecom/pull/1055 for more details.")
```
Like I said earlier, this slowdown isn't limited to GPU, CPU is also slower. It is limited to running with SVM though. So I'd change the check to 'using SVM and pocl >= 6' (we haven't tested pocl 7 yet).

majosm committed:
disable SVM on pocl6 where possible, otherwise warn

majosm commented:
@matthiasdiener Disabling SVM on non-NVIDIA seems OK as far as I can tell.  Timestep times:
|   | Mac (KS2D) | Lassen CPU (KS3D)|
| -------- | -------- | -------- |
| pocl5 | - | 1.51 s |
| pocl6+SVM | 0.548 s | 1.91 s |
| pocl6+CL buffer | 0.092 s | 1.42 s |

majosm commented:
One of the times that I ran CI here I got a failure in the `thermally-coupled-tpe` examples (output [here](https://gist.github.com/majosm/d719ed32482a87f5e398c592a6a01b7f)). I haven't been able to reproduce it since, so I'm going to let it go for now.

matthiasdiener commented:
> @matthiasdiener Disabling SVM on non-NVIDIA seems OK as far as I can tell. Timestep times:
> 
> Mac (KS2D)	Lassen CPU (KS3D)
> pocl5	-	1.51 s
> pocl6+SVM	0.548 s	1.91 s
> pocl6+CL buffer	0.092 s	1.42 s

Casual 6x speedup with CL buffers 🤷‍♂️
